server/server.go

```diff
@@ -540,3 +574,74 @@ func (s *Server) resolveAuthorizationModelID(ctx context.Context, store, modelID
 
 	return modelID, nil
 }
+
+// serverHealthChecker implements the HealthCheckManager interface for an OpenFGA server
+// specifically.
+type serverHealthChecker struct {
+	healthServer  *health.AuthlessHealthServer
+	openfgaServer *Server
+	serviceNames  map[string]struct{}
+}
+
+var _ health.HealthCheckManager = &serverHealthChecker{}
+
+// NewOpenFGAServerHealthChecker constructs a HealthCheckManager that can be used
+// to report the health status of the provided OpenFGA server.
+func NewOpenFGAServerHealthChecker(s *Server) health.HealthCheckManager {
+	return &serverHealthChecker{
+		healthServer:  health.NewAuthlessHealthServer(),
+		openfgaServer: s,
+		serviceNames:  map[string]struct{}{},
+	}
+}
+
+// RegisterService registers the provided serviceName with this server health checker and
+// sets it's serving status to 'NOT SERVING'
+func (s *serverHealthChecker) RegisterService(serviceName string) {
+	s.serviceNames[serviceName] = struct{}{}
+	s.healthServer.Server.SetServingStatus(serviceName, healthv1pb.HealthCheckResponse_NOT_SERVING)
+}
+
+// GetHealthServer returns the underlying health server managed by this health
+// checker.
+func (s *serverHealthChecker) GetHealthServer() *health.AuthlessHealthServer {
+	return s.healthServer
+}
+
+// Check reports whether the server managed by this server health checker is
+// ready to accept traffic.
+func (s *serverHealthChecker) Check(ctx context.Context) func() error {
+	return func() error {
+
+		backoffPolicy := backoff.NewExponentialBackOff()
+		backoffPolicy.MaxElapsedTime = 1 * time.Minute
+		ticker := backoff.NewTicker(backoffPolicy)
+		defer ticker.Stop()
+
+		// continuously monitor health status on a ticker interval
+		for {
```

Comment:
Implement an exponential backoff/retry to check the server readiness. If we exhaust our retry policy then we return an error and fatal error at the caller level. If we need to we can change this behavior, but it felt right to me at this time.